include/multipass/async_periodic_task.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef MULTIPASS_ASYNC_PERIODIC_TASK_H
+#define MULTIPASS_ASYNC_PERIODIC_TASK_H
+
+#include <multipass/logging/log.h>
+
+#include <chrono>
+#include <future>
+#include <string_view>
+
+#include <QTimer>
+
+namespace mpl = multipass::logging;
+
+namespace multipass::utils
+{
+// Because AsyncPeriodicTask needs to be instantiated as data member of a class to function properly, and Class template
+// argument deduction (CTAD) does not work for data member, so that means we can not deduced type data member from the
+// below code snippet.
+
+// template <typename Callable, typename...Args>
+// class AsyncPeriodicTask
+// {
+// public:
+//     using ReturnType = std::invoke_result_t<std::decay_t<Callable>, std::decay_t<Args>...>;
+//     AsyncPeriodicTask(Callable&& func, Args&&... args);
+//     ...
+// };
+
+// Therefore, the user has to specify the ReturnType as template argument while constructing it.
+
+template <typename ReturnType = void>
+class AsyncPeriodicTask
+{
+public:
+    template <typename Callable, typename... Args>
+    AsyncPeriodicTask(std::string_view launch_msg, std::chrono::milliseconds msec, Callable&& func, Args&&... args)
+    {
+        // Log in a side thread will cause some unit tests (like launch_warns_when_overcommitting_disk) to have data
+        // race on log, because the side thread log messes with the mock_logger. Because of that, we only allow the
+        // main thread log for now. That is why launch_msg parameter is here. A long-term solution would be a better
+        // separation of classes and mock the corresponding class function, so it will not log and mess with the
+        // mock_logger in the unit tests.
+
+        // TODO, remove the launch_msg parameter once we have better class separation.
+        mpl::log(mpl::Level::info, "async task", std::string(launch_msg));
+        future = std::async(std::launch::async, std::forward<Callable>(func), std::forward<Args>(args)...);
+        QObject::connect(&timer, &QTimer::timeout, [launch_msg, this, func, args...]() -> void {
+            // skip it if the previous one is still running
+            if (future.wait_for(std::chrono::seconds(0)) == std::future_status::ready)
+            {
+                mpl::log(mpl::Level::info, "async task", std::string(launch_msg));
```

Comment:
Hmm, I thought this should be seen even by the user when they look into the log, not just for us.

Reply:
I'm not sure if a user knowing that an asynchronous task started is something they should know about.  It seems more a debugging type thing, but perhaps others have an opinion.  @ricab, @luis4a0, et. al., any thoughts?

Reply:
Hmm, good question. I guess it's a little subjective... It depends on the task in question, on the frequency...

On the one hand, asynchronous tasks in general are an implementation detail. On the other, they probably involve a significant unit of work that the user may be interested in. But does it run every few seconds? We could parameterize the task class with how informative we want it to be, but that would feel like overkill here honestly. 

I would consider the concrete tasks this will be used for. Provided they are infrequent enough (15min sounds reasonable), and interesting enough (fetching image info arguably is), I would be fine with both `debug` and `info`. It can always be changed again if people complain.

Note: a period of 15min means 60*24/15 = 96 logs per day. IMO, that is perfectly within boundaries for a daemon like ours. It adds up to 34K/year or so, but that's just the nature of logging. Consumers should know that logs add up. Hopefully log facilities manage it by archiving, rotating, or something else.

Reply:
Hey! I agree, this should be `debug`. The user doesn't need to know when the manifests are being fetched.

Reply:
Thanks for all the inputs, I will change it to `debug`.